### PR TITLE
WIP: Displaying package parameters

### PR DESCRIPTION
**Pull request by fkz**
See also http://thread.gmane.org/gmane.linux.distributions.nixos/19714
Demo: http://schmitthenner.eu/nixos-homepage/nixos/packages.html

Not yet ready for review but I wanted to start collecting a TODO list at a central place.
# TODO for being able to publish
- [x] call the right patched nix version (instead of the hardcoded ~/... path)
- [x] revert the gunzip (which I need locally because I don't know how pregzipped files work with apache)
- [x] <del> use allow-unsafe-native-code-during-evaluation instead of patching nix (so we don't have to recompile nix) </del> not possible because we added comment (docstring) parsing to the parser of nix now
- [ ] make displaying stable and unstable versions more general
- [x] rebase
# Other nice additions
- [ ] show default value of optional parameters
- [x] investigate the "some error happened" packages
- [x] don't show failed packages (or mark them somehow)
- [ ] show more metadata like comments before parameters (or use more elaborate docstrings etc.)
- [ ] investigate what to do with more elaborate packages, e. g. `vim_configurable` that uses `composableDerivation`
- [x] maybe enable search for parameters
- [ ] make it more beginner friendly: show expression you should insert into `.nixpkgs/config.nix`

### CODE REVIEW

**Comment by domenkozar** [comment]
Very nice!

File: nixpkgs/packages.json-generation/addAttrs.nix [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+      forceCatch (error: [ ({ ${n + "(error)"} = { name = n; inherit error; meta = { longDescription = error.message; }; }; }) ])
+      (if lib.isDerivation x then
+        [(lib.setAttrByPath [n] (info x))]
+      else if lib.isAttrs x then
+        if x ? "recurseForDerivations" && x.recurseForDerivations then
+            lib.concatMap (name: mapRecursively "${n}.${name}" x.${name}) (lib.attrNames x)
+        else
+          []
+      else if lib.isList x then
+        [] # map (mapRecursively "${n}.list") x
```
Comment (moretea) [inline nixpkgs/packages.json-generation/addAttrs.nix]: Maybe we should raise an error, or actually recurse over it.

File: nixpkgs/packages.json-generation/nixpkgs.patch [hunk 2]
```diff
@@ -0,0 +1,27 @@
+diff --git a/lib/customisation.nix b/lib/customisation.nix
+index 5854954..57995a4 100644
+--- a/lib/customisation.nix
++++ b/lib/customisation.nix
+@@ -60,9 +60,13 @@ rec {
+         { override = newArgs: makeOverridable f (overrideWith newArgs);
+           overrideDerivation = fdrv:
+             makeOverridable (args: overrideDerivation (f args) fdrv) origArgs;
++          inherit origArgs;
```
Comment (moretea) [inline nixpkgs/packages.json-generation/nixpkgs.patch]: Maybe we shouldn't inherit, it's a bit dangerous. `_origArgs = origArgs`?

File: nixpkgs/packages.json-generation/nixpkgs.patch [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
++          _functionArgs = builtins.functionArgs f;
+         })
+       else if builtins.isFunction ff then
+         { override = newArgs: makeOverridable f (overrideWith newArgs);
++          inherit origArgs;
```
Comment (moretea) [inline nixpkgs/packages.json-generation/nixpkgs.patch]: Same here

**Comment by samueldr** [comment]
[triage]

Is this still a relevant feature to be developed for the website? The channel selection would be reimplemented in #209, the part about the package parameter isn't, but probably could be ported.

If the feature is desired, but the codebase in need for a rewrite, an issue for the feature request could be opened instead.

**Comment by garbas** [comment]
closing this due to inactivity, but if needed please reopen it.

**Pull request closed without merging**